### PR TITLE
Fix DSD PCM buffer accounting and bound retention

### PR DESCRIPTION
**Pull request by mwd102**
# What does this implement/fix?

This fixes a mismatch between the PCM format requested from FFmpeg and how the buffer counts it for DSF files.

For DSF reported as 352.8 kHz / 8-bit, MA requests 32-bit integer PCM but still counts the output as 8-bit. That has the effect of making three seconds of audio look like twelve seconds in the buffer, and a seek to 1.5 seconds actually lands at 0.375 seconds. The fix requests 32-bit float PCM and aligns the buffer calculations.

DFF/DST files with unknown codec metadata can also fall back to 16-bit PCM. Using f32 avoids losing precision before MA selects the final output format.

Based on review, to prevent buffers becoming too large, caps of 64/128/256 MiB per DSD PCM buffer for Minimal/Balanced/Maximum are implemented. With Balanced, that's ~47 seconds of stereo DSD64 or 23 seconds of DSD128. 

Additionally DFF detection implemented here now handles empty paths, multipart files, and local filenames with reserved chars. To respect upstream logic, if a provider explicitly supplies a decoded format, that takes priority over the filename.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue) — `bugfix`
- [ ] New feature (non-breaking change which adds functionality) — `new-feature`
- [ ] Enhancement to an existing feature — `enhancement`
- [ ] New music/player/metadata/plugin provider — `new-provider`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — `breaking-change`
- [ ] Refactor (no behaviour change) — `refactor`
- [ ] Documentation only — `documentation`
- [ ] Maintenance / chore — `maintenance`
- [ ] CI / workflow change — `ci`
- [ ] Dependencies bump — `dependencies`

## Checklist

- [x] The code change is tested and works locally.
- [x] `pre-commit run --all-files` passes.
- [x] `pytest` passes, and tests have been added/updated under `tests/` where applicable.
- [ ] For changes to shared models, the companion PR in `music-assistant/models` is linked.
- [ ] For changes affecting the UI, the companion PR in `music-assistant/frontend` is linked.
- [x] I have read and complied with the project's [AI Policy](https://github.com/music-assistant/.github/blob/main/AI_POLICY.md) for any AI-assisted contributions.
- [ ] I have [raised a PR against the documentation repository](https://github.com/music-assistant/music-assistant.io/blob/main/CONTRIBUTING.md) targeting the main or beta branch as appropriate.

### CODE REVIEW

File: music_assistant/helpers/audio.py [hunk 1]
```diff
@@ -755,6 +765,52 @@ def arriving_audio_format(streamdetails: StreamDetails) -> AudioFormat:
...
+    Return the PCM format FFmpeg emits after decoding the arriving audio.
+
+    FFprobe describes DSD in units of one byte per decoded time step: DSD64 is
+    reported as 352.8 kHz / 8-bit even though FFmpeg's DSD decoder emits planar
+    float samples. Normalize that decoder contract here so buffer byte accounting
+    and downstream bit-depth selection describe the bytes FFmpeg actually writes.
+
+    :param streamdetails: The stream whose decoded PCM format is required.
+    """
+    from music_assistant_models.media_items import AudioFormat  # noqa: PLC0415
```
Comment (marcelveldt) [inline music_assistant/helpers/audio.py]: why is the import inlined ?

Comment (mwd102) [reply music_assistant/helpers/audio.py]: Was following the convention from TYPE_CHECKING - assumed maybe there was a reason it wasn't typically loaded at runtime

File: music_assistant/helpers/audio.py [hunk 2]
```diff
@@ -755,6 +765,52 @@ def arriving_audio_format(streamdetails: StreamDetails) -> AudioFormat:
...
+        or audio_format.codec_type in DSD_CONTENT_TYPES
+    )
+
+
+def is_dsd_stream(streamdetails: StreamDetails) -> bool:
+    """Return whether a stream contains DSD, including DFF/DST local files."""
+    if is_dsd_audio_format(arriving_audio_format(streamdetails)):
+        return True
+    path = streamdetails.path
+    return isinstance(path, str) and urllib.parse.urlparse(path).path.lower().endswith(".dff")
```
[the remark this answers is not in the document]
Comment (mwd102) [reply music_assistant/helpers/audio.py]: Good catch, updated via commit 59026f695

decoded_audio_format now takes precedence and multipart paths are now handled.

**Comment by OzGav** [comment]
Please resolve the conversations as you fix the problem

File: music_assistant/helpers/audio.py [hunk 3]
```diff
@@ -755,6 +765,58 @@ def arriving_audio_format(streamdetails: StreamDetails) -> AudioFormat:
...
+    if is_dsd_audio_format(streamdetails.audio_format):
+        return True
+    path = streamdetails.path
+    if isinstance(path, str):
+        return urllib.parse.urlparse(path).path.lower().endswith(".dff")
+    return (
+        isinstance(path, list)
+        and bool(path)
+        and all(urllib.parse.urlparse(part.path).path.lower().endswith(".dff") for part in path)
+    )
```
Comment (MarvinSchenkel) [inline music_assistant/helpers/audio.py]: [CRITICAL] `urlparse()` treats `#` and `?` in local filenames as URL delimiters: `/music/Album #1/track.dff` becomes `/music/Album `, so DST/DFF files with unknown codec metadata still fall back to 16-bit PCM. Keep local paths intact and use the parsed path only for HTTP(S) URLs, for both single-file and multipart streams.

```suggestion
    path = streamdetails.path
    if not path:
        return False

    paths = [path] if isinstance(path, str) else [part.path for part in path]
    for file_path in paths:
        parsed = urllib.parse.urlparse(file_path)
        if parsed.scheme in ("http", "https"):
            file_path = parsed.path
        if not file_path.lower().endswith(".dff"):
            return False
    return True
```

Please cover local paths containing `#` and `?`, multipart equivalents, and HTTP URLs with query strings.

Comment (mwd102) [reply music_assistant/helpers/audio.py]: Addressed in  [`7fb78dd36`](https://github.com/music-assistant/server/pull/6177/commits/7fb78dd36) - small tweak to the loop to keep Ruff happy. Test cases added too.

**Comment by MarvinSchenkel** [comment]
Also looks like the tests are failing because the `is_dsd_stream` returns `True` for empty paths

**Review by MarvinSchenkel — CHANGES_REQUESTED** [review]

File: music_assistant/helpers/audio.py [hunk 4]
```diff
@@ -755,6 +765,60 @@ def arriving_audio_format(streamdetails: StreamDetails) -> AudioFormat:
...
+    :param streamdetails: The stream whose decoded PCM format is required.
+    """
+    arriving = arriving_audio_format(streamdetails)
+    if is_dsd_stream(streamdetails):
+        return AudioFormat(
+            content_type=ContentType.PCM_F32LE,
+            codec_type=ContentType.PCM_F32LE,
+            sample_rate=arriving.sample_rate,
+            bit_depth=32,
+            channels=min(arriving.channels, 2),
```
[the remark this answers is not in the document]
Comment (mwd102) [reply music_assistant/helpers/audio.py:819]: Agree, buffer is bigger than needs to be - fixed  by capping retained DSD PCM at 64/128/256 MiB per buffer for Minimal/Balanced/Maximum. Tests passing locally, also verified seeking and buffers behaves as expected with real DSF and DST128  files.

[the remark this answers is not in the document]
Comment (mwd102) [reply music_assistant/helpers/audio.py:819]: I went back and checked the timing issue too. Made mistake in description, meant S32, not F32. But the buffer was still counting it as a byte, so the timing bug was real. Real DSD stream testing resulted in  three seconds of audio counted as twelve.

[the remark this answers is not in the document]
Comment (mwd102) [reply music_assistant/helpers/audio.py:819]: Description updated to reflect change and be clearer on one issue based on copilot findings. Added regression test to more clearly demonstrate timing issue.

**Review by MarvinSchenkel — APPROVED** [review]
Thanks @mwd102
